### PR TITLE
Handle currentUserId in repo layer + View Upcoming Events

### PR DESCRIPTION
**Pull request by Nath-Assa**
### Key Changes:

**User Repository Updates:**
- Implemented getCurrentUserId() in IUserRepository, MockUserRepository, and FirebaseUserRepository.
- Added unit tests for the new getCurrentUserId() method to ensure correct behavior.

**ViewModel Refinements:**
- Transitioned EventOverviewViewModel state properties from var to val for immutability and introduced userLoggedIn property
- Introduced init() and checkUserLoginStatus() to manage user session state based on repository changes.
- Enhanced event handling with two new methods:
   - `setFilterDialogOpen(isOpen: Boolean)`: Manages the visibility of the filter dialog.
   - `onSearchQueryChange(newQuery: String)`: Updates the search query state.
- Expanded EventOverviewViewModelTest to include tests for the new methods and changes.

**UI Integration:**
- Adjusted LaunchedEffect in Home.kt to trigger getEvents() unconditionally on component composition instead of reacting to eventList changes.
- Updated Home.kt to display upcoming events dynamically based on user session status.
- Implemented UI tests to validate the frontend integration with the newly adjusted view models.

### CODE REVIEW

File: app/src/main/java/com/github/se/eventradar/viewmodel/EventsOverviewViewModel.kt
Comment (Nath-Assa) [inline app/src/main/java/com/github/se/eventradar/viewmodel/EventsOverviewViewModel.kt:142]: @christinehahaa Take a look at this please.

File: app/src/main/java/com/github/se/eventradar/model/repository/user/MockUserRepository.kt
Comment (wisdomseeker2001) [inline app/src/main/java/com/github/se/eventradar/model/repository/user/MockUserRepository.kt:8]: This could be made private

File: app/src/main/java/com/github/se/eventradar/ui/home/Home.kt
Comment (wisdomseeker2001) [inline app/src/main/java/com/github/se/eventradar/ui/home/Home.kt:138]: good respect of MVVM model. Nice work

File: app/src/main/java/com/github/se/eventradar/viewmodel/EventsOverviewViewModel.kt
Comment (wisdomseeker2001) [inline app/src/main/java/com/github/se/eventradar/viewmodel/EventsOverviewViewModel.kt:51]: have you thought of perhaps putting this in a some sort of loop to treaty fetching the events? perhaps firebase bugs at the instance u attempted, I personally think it is worth reattempting.

File: app/src/main/java/com/github/se/eventradar/viewmodel/EventsOverviewViewModel.kt
Comment (wisdomseeker2001) [inline app/src/main/java/com/github/se/eventradar/viewmodel/EventsOverviewViewModel.kt:81]: did not quite get why events.data is there 2x

File: app/src/main/java/com/github/se/eventradar/viewmodel/EventsOverviewViewModel.kt
Comment (wisdomseeker2001) [inline app/src/main/java/com/github/se/eventradar/viewmodel/EventsOverviewViewModel.kt:101]: isn't a user always need to be logged In for our V1? This is defo good to forsee updates of app in long run

File: app/src/main/java/com/github/se/eventradar/viewmodel/EventsOverviewViewModel.kt
Comment (wisdomseeker2001) [inline app/src/main/java/com/github/se/eventradar/viewmodel/EventsOverviewViewModel.kt:1]: only thing I could suggest u add is some ENUM that enums different states the viewmodel goes through. e.g. Enum State { FbFetchErr, FbUpdateErr ...} and in your tests you can then test that a state flow observing this enum changed In the predicted sequence of states.

Assume x is a Stateflow, you can do x.take(n).toList where n is number of stats you expected it to change to in the test you called and then you can assert it was = to Listen (State1, state2 ,...) This is good practice as It will ensure that the flow of code is exactly as you expected

File: app/src/main/java/com/github/se/eventradar/model/repository/user/MockUserRepository.kt
Comment (wisdomseeker2001) [inline app/src/main/java/com/github/se/eventradar/model/repository/user/MockUserRepository.kt:1]: overall really good work, this is super useful for testing now that you can orchestrate custom UserIds

File: app/src/main/java/com/github/se/eventradar/viewmodel/EventsOverviewViewModel.kt
Comment (nikhen-s) [inline app/src/main/java/com/github/se/eventradar/viewmodel/EventsOverviewViewModel.kt:101]: I think this function right now looks more like updating the user's log in status. It would be nice upon Resource.Failure to Log a statement saying that no user is logged in for easy testing too.

File: app/src/main/java/com/github/se/eventradar/viewmodel/EventsOverviewViewModel.kt
Comment (nikhen-s) [inline app/src/main/java/com/github/se/eventradar/viewmodel/EventsOverviewViewModel.kt:117]: Just fyi, you don't need to change this because I've already told christine, but credits to [Jalen's comment](https://github.com/swent-sp-2024-event-radar/event-radar-app/pull/130#discussion_r1585952971) - this doesn't need to take in the boolean. Whenever it's called, we know the new value will be the opposite of what it is now.

File: app/src/test/java/com/github/se/eventradar/EventOverviewViewModelTest.kt
Comment (nikhen-s) [inline app/src/test/java/com/github/se/eventradar/EventOverviewViewModelTest.kt:134]: Is that a reason why you casted this as MockUserRepository? I think it's been setup as `userRepository = MockUserRepository()` @Before the tests run, curious to hear your reason.

File: app/src/test/java/com/github/se/eventradar/MockUserRepositoryUnitTest.kt
Comment (nikhen-s) [inline app/src/test/java/com/github/se/eventradar/MockUserRepositoryUnitTest.kt:1]: Tests LGTM :)

File: app/src/main/java/com/github/se/eventradar/ui/home/Home.kt
Comment (nikhen-s) [inline app/src/main/java/com/github/se/eventradar/ui/home/Home.kt:65]: concern: If getUpcomingEvents() updates the uiState.eventList, and getEvents() is triggered whenever uiState.eventList changes, there's a possibility of an unintended recursion or multiple calls to getEvents().

**Review by nikhen-s — CHANGES_REQUESTED** [review]
Some questions and nits, but nice Job on the getUserId function & tests. Upon trying to run the app on my device, clicking on the `upcomingEvents` tab still does not change the tab (it stays in Browse tab), try to look into that and add some UI tests in androidTest folder for the UpcomingEvents part (and switching tabs). Don't forget to check [sonar cloud](https://sonarcloud.io/project/issues?resolved=false&pullRequest=184&id=swent-sp-2024-event-radar_event-radar-app&open=AY80hbqL0-tUbcqyDyHT) to address the maintainability issues too.

File: app/src/main/java/com/github/se/eventradar/ui/home/Home.kt
Comment (christinehahaa) [inline app/src/main/java/com/github/se/eventradar/ui/home/Home.kt:135]: Search and filter was my job, so I have made these changes already in my incoming PR. But since you've created tests for them already, let's keep them here. 

I will update your functions/tests related to search and filter with my work where necessary, since I have already tested on my end to ensure it works.

File: app/src/main/java/com/github/se/eventradar/viewmodel/EventsOverviewViewModel.kt
Comment (christinehahaa) [reply app/src/main/java/com/github/se/eventradar/viewmodel/EventsOverviewViewModel.kt:81]: EventList contains allEvents and FilteredEvents. First one is allEvents, second is FilteredEvents. When data is first fetched, FilteredEvents =  allEvents.

File: app/src/main/java/com/github/se/eventradar/viewmodel/EventsOverviewViewModel.kt
Comment (christinehahaa) [inline app/src/main/java/com/github/se/eventradar/viewmodel/EventsOverviewViewModel.kt:122]: I will edit these functions with my work if needed, once I push my PR.

File: app/src/main/java/com/github/se/eventradar/viewmodel/EventsOverviewViewModel.kt
Comment (christinehahaa) [inline app/src/main/java/com/github/se/eventradar/viewmodel/EventsOverviewViewModel.kt:136]: Remove this TO DO, I've implemented already. Thank you for adding them tho :)

File: app/src/main/java/com/github/se/eventradar/viewmodel/EventsOverviewViewModel.kt
Comment (christinehahaa) [reply app/src/main/java/com/github/se/eventradar/viewmodel/EventsOverviewViewModel.kt:143]: Looks good, let's keep this in your PR since you've done tests for it already. That way the work is associated to you. 

I will change all my other UiState values I added to be val and not var.

File: app/src/test/java/com/github/se/eventradar/EventOverviewViewModelTest.kt
Comment (christinehahaa) [inline app/src/test/java/com/github/se/eventradar/EventOverviewViewModelTest.kt:240]: Thanks for adding these :)

**Review by christinehahaa — COMMENTED** [review]
Some comments regarding code overlap with my PR, but otherwise I tested on my machine and looks good overall. Line coverage just needs to be boosted to above 80% before merging into main.

File: app/src/main/java/com/github/se/eventradar/viewmodel/EventsOverviewViewModel.kt
Comment (Nath-Assa) [reply app/src/main/java/com/github/se/eventradar/viewmodel/EventsOverviewViewModel.kt:51]: I believe this works fine. Don't think changing this would be time-efficient.

File: app/src/main/java/com/github/se/eventradar/viewmodel/EventsOverviewViewModel.kt
Comment (Nath-Assa) [reply app/src/main/java/com/github/se/eventradar/viewmodel/EventsOverviewViewModel.kt:101]: Yes, a User should always be logged in. This will be useful long term.

File: app/src/main/java/com/github/se/eventradar/viewmodel/EventsOverviewViewModel.kt
Comment (Nath-Assa) [reply app/src/main/java/com/github/se/eventradar/viewmodel/EventsOverviewViewModel.kt:1]: Would be nice indeed. Something to look into during next sprint definitely.

File: app/src/main/java/com/github/se/eventradar/ui/home/Home.kt
Comment (Nath-Assa) [reply app/src/main/java/com/github/se/eventradar/ui/home/Home.kt:201]: yes! this is why the upcoming tab wasn't working. Will remove it.

File: app/src/main/java/com/github/se/eventradar/ui/home/Home.kt
Comment (Nath-Assa) [reply app/src/main/java/com/github/se/eventradar/ui/home/Home.kt:182]: In the current implementation, the scenario where a user attempts to access 'Upcoming Events' without being logged in should indeed be impossible due to the app's navigation guards. However, including this check and displaying a relevant message ('Please log in') serves as an additional layer of error handling.

Though ideally, users should never see this message under normal circumstances, it acts as a fail-safe to prompt users towards corrective action (like logging in) if they somehow navigate to this state.

File: app/src/main/java/com/github/se/eventradar/viewmodel/EventsOverviewViewModel.kt
Comment (Nath-Assa) [reply app/src/main/java/com/github/se/eventradar/viewmodel/EventsOverviewViewModel.kt:117]: I believe Christine said that I would be keeping this method in this PR. Will edit.

File: app/src/main/java/com/github/se/eventradar/viewmodel/EventsOverviewViewModel.kt
Comment (Nath-Assa) [reply app/src/main/java/com/github/se/eventradar/viewmodel/EventsOverviewViewModel.kt:101]: This function invokes getCurrentId(), which already handles logging a message if a user is not signed in. Including an additional log message here would be redundant.

File: app/src/test/java/com/github/se/eventradar/EventOverviewViewModelTest.kt
Comment (Nath-Assa) [reply app/src/test/java/com/github/se/eventradar/EventOverviewViewModelTest.kt:134]: Yes, the method updateCurrentUserID() is exclusive to the MockUserRepository and is not defined in the IUserRepository interface. As a result, casting is necessary to access this method, since it would not be available through the interface otherwise.

File: app/src/main/java/com/github/se/eventradar/ui/home/Home.kt
Comment (Nath-Assa) [reply app/src/main/java/com/github/se/eventradar/ui/home/Home.kt:65]: Yes. Will update this.

**Comment by Nath-Assa** [comment]
Still need to: 
- Add extra UI tests
- Add tests for getCurrentUserId() in FirebaseUserRepository
- Complete the upgrade of HostedEventsViewModel to align with the current user ID handling.

**Review by christinehahaa — APPROVED** [review]
Upcoming events tab is working now -  nice job!

For next time - it appears there is a lot of changes being made in this PR, some related, some unrelated to each other. I would split into multiple PRs so they can be pushed into main faster next time :)

**Comment by Nath-Assa** [comment]
**Note:** The implementations of EventList and EventMap in Home.kt may seem to be near-duplicate code. However, each usage is contextually unique with specific test tags tailored to different UI scenarios. While extracting these into private functions could reduce code duplication superficially, it would also add complexity. After careful consideration, it was decided to keep these implementations separate to maintain code clarity and simplicity.

File: app/src/main/java/com/github/se/eventradar/ui/home/Home.kt
Comment (nikhen-s) [inline app/src/main/java/com/github/se/eventradar/ui/home/Home.kt:175]: Do you think we should still show Map View even if there are no Upcoming Events available? The map will be empty, because i think this might be what users would expect upon clicking the MapView, but we can ask others how they think about this too.

File: app/src/main/java/com/github/se/eventradar/ui/home/Home.kt
Comment (nikhen-s) [inline app/src/main/java/com/github/se/eventradar/ui/home/Home.kt:60]: nit: maybe it'a better to initialize this in the viewModel using `init{}` for separation of concerns?

File: app/src/main/java/com/github/se/eventradar/viewmodel/EventsOverviewViewModel.kt
Comment (nikhen-s) [inline app/src/main/java/com/github/se/eventradar/viewmodel/EventsOverviewViewModel.kt:49]: Is there a particular reason you removed uid here?

File: app/src/main/java/com/github/se/eventradar/ui/home/Home.kt
Comment (nikhen-s) [inline app/src/main/java/com/github/se/eventradar/ui/home/Home.kt:146]: small nit: instead of if (getTabIndexFromTabEnum(uiState.tab) == 0), maybe uiState.tab == Tab.Browsing would be more readable

File: app/src/main/java/com/github/se/eventradar/viewmodel/EventsOverviewViewModel.kt
Comment (nikhen-s) [inline app/src/main/java/com/github/se/eventradar/viewmodel/EventsOverviewViewModel.kt:69]: small nit: i know that we separated hosting viewmodel and eventsOverviewViewModel, but maybe getUserUpcomingEvents is more clear so it does not get confused with getUserHostedEvents.

**Review by nikhen-s — CHANGES_REQUESTED** [review]
Overall looks great, but just need to address 2 things and after it LGTM:

1. Try to fix this [maintainability issue](https://sonarcloud.io/project/issues?resolved=false&pullRequest=184&id=swent-sp-2024-event-radar_event-radar-app&open=AY80hbqL0-tUbcqyDyHT&tab=code) in SonarCloud to help increase the code quality. (It would also help to reduce the code dupliaction, suggestion: I put the if statements through ChatGPT and it helped suggest some ways to reuse the code & reduce the duplication)
2. (Smaller question) Should we display the MapView even if there are no Upcoming Events?

Other comments are just nits

File: app/src/main/java/com/github/se/eventradar/viewmodel/EventsOverviewViewModel.kt
Comment (Nath-Assa) [reply app/src/main/java/com/github/se/eventradar/viewmodel/EventsOverviewViewModel.kt:49]: This change is part of a broader update where user identification is handled directly within the repository layer, specifically using the new getCurrentUserId() method from the IUserRepository interface.

File: app/src/main/java/com/github/se/eventradar/ui/home/Home.kt
Comment (Nath-Assa) [reply app/src/main/java/com/github/se/eventradar/ui/home/Home.kt:60]: LaunchedEffect with Unit as the key ensures that getEvents() is called once when the composable enters the composition and not again unless the composable is recomposed from scratch. 
Triggering data loading in the ViewModel's init{} block can lead to premature requests, misaligned lifecycles, and unnecessary resource consumption.

File: app/src/main/java/com/github/se/eventradar/ui/home/Home.kt
Comment (Nath-Assa) [reply app/src/main/java/com/github/se/eventradar/ui/home/Home.kt:175]: Good point! I agree this deserves further discussion with the team to ensure it aligns with user expectations. We can consider making adjustments in the next sprint based on our findings.

**Review by nikhen-s — APPROVED** [review]
LGTM

**Comment by Nath-Assa** [comment]
Closes #107 and #183

**Comment by Nath-Assa** [comment]
Note: Last 5 commits can be ignored was trying to get the CI to work.

**Comment by Nath-Assa** [comment]
**Detected small bug:**  When you are on Upcoming tab and press on the nav bar to navigate to another page, then come back to home, Upcoming tab shows all events(instead of upcoming). This is caused by:
`LaunchedEffect(Unit) { viewModel.getEvents() }`
Will fix this in another PR.

**Pull request merged**
